src/ray/core_worker/lessor_picker.h

```diff
@@ -0,0 +1,77 @@
+// Copyright 2017 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include "absl/base/thread_annotations.h"
+#include "absl/container/flat_hash_map.h"
+#include "absl/container/flat_hash_set.h"
+#include "ray/gcs/redis_gcs_client.h"
+
+namespace ray {
+
+struct LocalityData {
+  uint64_t object_size;
+  absl::flat_hash_set<NodeID> nodes_containing_object;
+};
+
+/// Interface for providers of locality data to the lessor picker.
+class LocalityDataProviderInterface {
+ public:
+  virtual absl::optional<LocalityData> GetLocalityData(const ObjectID &object_id) = 0;
+
+  virtual ~LocalityDataProviderInterface() {}
+};
+
+/// Interface for mocking the lessor picker.
+class LessorPickerInterface {
+ public:
+  /// Get the address of the best lessor node for the provided task.
+  virtual absl::optional<rpc::Address> GetBestNodeAddrForTask(
+      const TaskSpecification &spec) = 0;
+
+  virtual ~LessorPickerInterface() {}
+};
+
+typedef std::function<rpc::Address(const NodeID &node_id)> NodeAddrFactory;
+
+/// Class used by the core worker to pick the best lessor to request a lease from.
+/// This class is not thread safe.
+class LessorPicker : public LessorPickerInterface {
+ public:
+  LessorPicker(std::shared_ptr<LocalityDataProviderInterface> locality_data_provider,
+               NodeAddrFactory node_addr_factory)
+      : locality_data_provider_(locality_data_provider),
+        node_addr_factory_(node_addr_factory) {}
+
+  ~LessorPicker() {}
+
+  /// Get the address of the best lessor node for the provided task.
+  absl::optional<rpc::Address> GetBestNodeAddrForTask(const TaskSpecification &spec);
+
+  /// Get the best lessor node for the provided task.
+  absl::optional<NodeID> GetBestNodeForTask(const TaskSpecification &spec);
+
+  /// Get the best lessor node for the provided objects.
+  absl::optional<NodeID> GetBestNodeForObjects(const std::vector<ObjectID> &object_ids);
+
+ private:
+  /// Provider of locality data that will be used in choosing the best lessor.
+  std::shared_ptr<LocalityDataProviderInterface> locality_data_provider_;
+
+  /// Factory for building node RPC addresses given a NodeID.
+  NodeAddrFactory node_addr_factory_;
```

Comment:
Three other options that I have played around with:

1. Provide the GCS client to the `LessorPicker`, and move the factory function logic into `GetBestNodeAddrForTask`.
2. Provide the GCS client to the `CoreWorkerDirectTaskSubmitter`, change the interface of `LessorPickerInterface` to `GetBestNodeForTask` (removing `GetBestNodeAddrForTask` from `LessorPicker`), and move the factory function logic to the call site of `GetBestNodeForTask` in the direct task submitter.
3. Provide this factory function to the `CoreWorkerDirectTaskSubmitter` and change the interface of `LessorPickerInterface` to `GetBestNodeForTask` (removing `GetBestNodeAddrForTask` from `LessorPicker`).

I think that (1) and (2) expose an unnecessary amount of the GCS API surface to `LessorPicker` and `CoreWorkerDirectTaskSubmitter`, respectively, and is harder to test (would have to mock the `GcsClient` in the `LessorPicker`/`CoreWorkerDirectTaskSubmitter` tests), so I went with the factory injection pattern. (3) decouples all GCS/RPC address considerations from the `LessorPicker` (it would only deal with `NodeID`s) and doesn't add a lot of testing complexity, at the cost of moving the `NodeID` --> `rpc::Address` conversion up one abstraction level. I think that (3) might be a better route?